src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs

```diff
@@ -2822,6 +2829,67 @@ void EmitAtomicSingleCharLoop(RegexNode node)
                     }
                     Stloc(iterationLocal);
                 }
+                else if (node.Type == RegexNode.Setloopatomic &&
+                    maxIterations == int.MaxValue &&
+                    (numSetChars = RegexCharClass.GetSetChars(node.Str!, setChars)) > 1 &&
```

Comment:
I'm not understanding how we know that the set only contains 2 or 3 chars. Why won't we get into this code if there are 4 chars in the set?

Reply:
setChars is a span of only 3 elements, so it can't be more than that.